internal/charm/charmdir.go

```diff
@@ -4,351 +4,128 @@
-	} else {
-		b.config, err = ReadConfig(reader)
-		_ = reader.Close()
-		if err != nil {
-			return nil, errors.Annotatef(err, `parsing "config.yaml" file`)
-		}
+		return nil, errors.Errorf(`reading "manifest.yaml" file: %w`, err)
 	}
+	defer reader.Close()
 
-	if b.actions, err = getActions(
-		b.meta.Name,
-		func(file string) (io.ReadCloser, error) {
-			return os.Open(b.join(file))
-		},
-		func(err error) bool {
-			_, ok := err.(*os.PathError)
-			return ok
-		},
-	); err != nil {
-		return nil, err
+	manifest, err := ReadManifest(reader)
+	if err != nil {
+		return nil, errors.Errorf(`parsing "manifest.yaml" file: %w`, err)
 	}
 
-	if reader, err = os.Open(b.join("revision")); err == nil {
-		_, err = fmt.Fscan(reader, &b.revision)
-		_ = reader.Close()
-		if err != nil {
-			return nil, errors.New("invalid revision file")
-		}
-	}
+	return manifest, nil
+}
 
-	reader, err = os.Open(b.join("lxd-profile.yaml"))
+// ReadCharmDirConfig reads and parses the config file for a charm directory.
+func ReadCharmDirConfig(path string) (*Config, error) {
+	reader, err := os.Open(filepath.Join(path, "config.yaml"))
 	if _, ok := err.(*os.PathError); ok {
-		b.lxdProfile = NewLXDProfile()
+		return nil, errors.Errorf("config.yaml: %w", FileNotFound)
 	} else if err != nil {
-		return nil, errors.Annotatef(err, `reading "lxd-profile.yaml" file`)
-	} else {
-		b.lxdProfile, err = ReadLXDProfile(reader)
-		_ = reader.Close()
-		if err != nil {
-			return nil, errors.Annotatef(err, `parsing "lxd-profile.yaml" file`)
-		}
+		return nil, errors.Errorf(`reading "config.yaml" file: %w`, err)
 	}
+	defer reader.Close()
 
-	reader, err = os.Open(b.join("version"))
+	config, err := ReadConfig(reader)
 	if err != nil {
-		if _, ok := err.(*os.PathError); !ok {
-			return nil, errors.Annotatef(err, `reading "version" file`)
-		}
-	} else {
-		b.version, err = readVersion(reader)
-		_ = reader.Close()
-		if err != nil {
-			return nil, errors.Annotatef(err, `parsing "version" file`)
-		}
+		return nil, errors.Errorf(`parsing "config.yaml" file: %w`, err)
 	}
 
-	return b, nil
-}
-
-// join builds a path rooted at the charm's expanded directory
-// path and the extra path components provided.
-func (dir *CharmDir) join(parts ...string) string {
-	parts = append([]string{dir.Path}, parts...)
-	return filepath.Join(parts...)
+	return config, nil
 }
 
-// resolveSymlinkedRoot returns the target destination of a
-// charm root directory if the root directory is a symlink.
-func resolveSymlinkedRoot(rootPath string) (string, error) {
-	info, err := os.Lstat(rootPath)
-	if err == nil && info.Mode()&os.ModeSymlink != 0 {
-		rootPath, err = filepath.EvalSymlinks(rootPath)
-		if err != nil {
-			return "", fmt.Errorf("cannot read path symlink at %q: %v", rootPath, err)
-		}
+// ReadCharmDirActions reads and parses the actions file for a charm directory.
+func ReadCharmDirActions(charmName string, path string) (*Actions, error) {
```

Comment:
**suggestion:** that's really boring functions that could have been written once.

```go
func readCharmDirFile(file, charmName, path string, parser func(charmName string, io.Reader)(*Actions, error) ) (*Actions, error) {
	reader, err := os.Open(filepath.Join(path, file))
	if _, ok := err.(*os.PathError); ok {
		return nil, errors.Errorf("%s: %w", file, FileNotFound)
	} else if err != nil {
		return nil, errors.Errorf(`reading %q file: %w`,file, err)
	}
	defer reader.Close()

	actions, err := parser(charmName, reader)
	if err != nil {
		return nil, errors.Errorf(`parsing %q file: %w`, file, err)
	}
	return actions, nil
}

func ReadCharmDirActions(charmName; path string) (*Actions, error) {
   return readCharmDirFile("actions.yaml", charmName, path, ReadActionsYaml)
}
```

Reply:
I've gone somewhere in the middle because unfortunately `ReadCharmDirActions` is slightly different. I also don't like passing through functions unless we have to

Reply:
Actually on second thought, I don't think it's worth it. There are too many minor exceptions I don't think this would actually save us much

Reply:
Actually on second thought, I don't think it's worth it. There are too many minor exceptions I don't think this would actually save us much